### DISCUSSION TITLE
replace java.util.logging with slf4j

### DISCUSSION THREAD

**Discussion by vforchi** [nom-tam-fits/nom-tam-fits#203] [Ideas]
I am using your library in a number of applications, and I would like to have only one logging system.
Have you considered moving from java.util.logging to slf4j? That would make integration with e.g. Spring Boot applications much simpler.

**Comment by Zlika** [comment]
Hi. You can use jul-to-slf4j bridge as a workaround (cf. https://www.slf4j.org/legacy.html).

**Comment by vforchi** [comment]
Yes, but you can do that only if you know you have that dependency. If, for example, you are using a library that has nom.tam.fits as a dependency you may never notice that you are losing log messages, because they are not going to the same place as the rest.

**Comment by Zlika** [comment]
Maybe the solution would be to remove logging. I don't think this kind of library should log at all.
I made a quick search in the code base: lots of classes declare a logger, but I wasn't able to find a single one effectively using it!

**Comment by attipaci** [comment]
I tend to agree with @Zlika. I'm not sure the library should be logging. It should throw exceptions as appropriate rather than logger warnings. (This has been frustrating me a bit also). Let's not do anything rash though. We should carefully think through the pros and cons of removing logging. Especially, if this could break backward compatibility in any way. For example, if there could be a use case that relies on the logging feature to be present (I would hope not)... In any case, if you have interest in following up, I'll be happy to stay engaged. :-)

**Comment by vforchi** [comment]
If you throw exceptions you can't have debug information, so you would be more limited.
I really hope nobody is relying on the logging feature for their application to work, but you never know.
If you agree to use slf4j I would be happy to make a pull request.

**Comment by Zlika** [comment]
You can have all the debug info you want in the exception (message, backtrace...). Libraries should not use loggers. It's the job of the main application to log.

**Comment by vforchi** [comment]
I don't agree: exceptions are heavy and should be, as the name suggests, exceptions.
If I want to know how much time a certain function spends in its subroutines how would this work? And how can I control which exceptions are thrown and which are not?
As for libraries using loggers, see e.g. spring.

**Comment by Zlika** [comment]
Spring is a framework, not a library. It brings lots of constraints that a mere library should not. This issue is an example of the problems you can have when a lib uses loggers. Take as an example the jdk itself, it does not use loggers. From my point of view, using loggers in a library is an anti-pattern. Replacing jul by slf4j will only move the problem. The root problem is that a library does know the context of the applications that will use it, and so should not impose things like the logging framework (or even the logging facade) and should not decide what is relevant to log for the application.

**Comment by vforchi** [comment]
If you use slf4j the library does not decide what to log, the configuration file of the application does.

back to my example: if there are performance problems in some parts of the library, how should I debug it with exceptions?

**Comment by Zlika** [comment]
With a profiler, a debugger, a test case... Anyway, not with a logger.

**Comment by attipaci** [comment]
I second the opinion, that I don't think there is a strong case for keeping logging in the library.

I think part of the use case for the logger -- to select what errors we care for, and what we take a pass on -- could be done equally with exceptions, and corresponding settings (e.g. in `FitsFactory.FitsSettings`) that determine how tolerant we want to be to non-critical issues. If the setting is to tolerate specific problems, you go silently about it. And, if you want to know of minor issues, you change the setting and get an exception thrown... The trick is to figure out what settings and how many we want to control the generation of non-critical exceptions. My guess is at most a handful (like logger levels), and possibly all controlled via one routine (e.g. `static void Fits.setExceptionLevel(int or enum)`), which can take values that correspond to "none", "permissive", "strict", or "all" (for example).... What do you think of that as an idea to move forward with?

**Comment by vforchi** [comment]
I'm not sure how this would work: will the library throw an Exception in the middle of a routine (and fail) if I set it up in debug mode? And how would these Exceptions look like? I imagine this would make the library much more cumbersome to use, if somebody has to catch all these exceptions, and I personally would just disable them (hoping they are RuntimeExceptions.
This would also nullify the advantage of using a framework like Spring Boot, where one can change the logging level of specific classes or packages at runtime, provided they use slf4j.
I can understand the desire not to log anything, and I would be okay with that, but throwing Exceptions does not sound like a good solution to me.

**Comment by attipaci** [comment]
And yet, the entire core Java is designed around exceptions and nothing else. It's hard for me to see why would it not work for our limited-scope library, when it seems to work just fine for the entire Java framework...

Again, what I am proposing is that you can control whether or not an exception is thrown for example if the FITS violates some standard. If you don't want to deal with the exception, you simply turn it off completely (which in fact should be the default, by the way). (And if you don't care, why do you even want to log it?). Conversely, if you do care, e.g. because your application needs to conform to FITS standards strictly, then logging is not the right choice either. In that case you want the punch-in-the-face of the exception to call attention to a problem in your application that you want to / have to fix...

And, yes to using `RuntimeException`s in general to indicate non-critical issues when these are enabled. Specifically, we would probably want a `FitsStandardException` (derived probably from `IllegalStateException`), whenever we encounter something that does not conform to proper FITS standard (but we may let it pass). Also, I'd advocate the default exception level to be very permissive, that is not to throw an exception by default when encountering a non-critical issue (such as a malformed header entry). But, the developer can then enable runtime exceptions, as they need them, e.g. to create FITS files that do not deviate from the standards...

**Comment by vforchi** [comment]
If we limit this to failures regarding the FITS standard I'm ok with it, I don't understand how this would work for debugging.
On a side note, I'm not a big fan of the global configurations in `FitsFactory`, and I would rather have them as property of the `Fits` object.

**Comment by attipaci** [comment]
I agree with you on the global conf. If it is moved, we'll have to make sure to support the old public API for back compatibility...

Back to the logging, I'm not dead set against leaving the door open for logging support in some way (e.g. for debugging), but if we do that, I think we should follow two rules:

1. No logging is enabled by default. (Logging is an option that a developer can enable).
2. We don't pre-determine what logging facility the developer can use (and the default is none). For example, logging in the FITS library is performed via a designated new `Logging` interface (probably under `nom.tam.util`, or `nom.tam.util.logging`), which provides some set of basic methods, probably similar to that of `Logger` itself. This interface can be then implemented by adapter classes (e.g. `LoggerAdapter`, `SLF4JApdapter` which map those methods to the implementations of `Logger`, __slf4j__, respectively). This way, we don't pre-determine what logging facility the developer can use, and they can write their own adapters to any other logging facility in any way they see fit...

And logging or not, I'd still throw exceptions if strict compliance is desired and the library is making choices that deviate from the standard...

**Comment by mbtaylor** [comment]
I don't have a firm opinion on whether logging is a good idea or not in nom.tam.fits, but I would be against adding a new required dependency outside of the J2SE (slf4j) to the package.  No objection to the pluggable adapter pattern.
